sdk/azidentity/cache/internal/aescbc/aescbc.go

```diff
@@ -0,0 +1,118 @@
+//go:build go1.18
+// +build go1.18
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package aescbc
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"encoding/binary"
+	"errors"
+	"hash"
+)
+
+// AESCBCHMACSHA2 implements AES_CBC_HMAC_SHA2 as defined in https://tools.ietf.org/html/rfc7518#section-5.2.2
+type AESCBCHMACSHA2 struct {
+	Alg            string
+	encKey, macKey []byte
+	hasher         func() hash.Hash
+	tLen           int
+}
+
+type EncryptResult struct {
+	Ciphertext, Tag []byte
+}
+
+// NewAES128CBCHMACSHA256 returns an implementation of AES_128_CBC_HMAC_SHA_256
+// (https://tools.ietf.org/html/rfc7518#section-5.2.3)
+func NewAES128CBCHMACSHA256(key []byte) (*AESCBCHMACSHA2, error) {
+	if len(key) != 32 {
+		return nil, errors.New("key must be 32 bytes")
+	}
+	cp := make([]byte, 32)
+	copy(cp, key)
+	return newAESCBCHMACSHA2("A128CBC-HS256", cp, crypto.SHA256.New)
+}
+
+func newAESCBCHMACSHA2(alg string, k []byte, hasher func() hash.Hash) (*AESCBCHMACSHA2, error) {
+	return &AESCBCHMACSHA2{
+		Alg:    alg,
+		encKey: k[len(k)/2:],
+		hasher: hasher,
+		macKey: k[:len(k)/2],
+		tLen:   len(k) / 2,
+	}, nil
+}
+
+func (a *AESCBCHMACSHA2) Decrypt(iv, ciphertext, additionalData, tag []byte) ([]byte, error) {
+	expected := a.tag(iv, ciphertext, additionalData)
+	if !hmac.Equal(tag, expected) {
+		return nil, errors.New("decryption failed")
```

Comment:
Does this error get surfaced outside of `azidentity`?  If so, perhaps a bit more descriptive message?

Reply:
Comment applies to a few other places.

Reply:
This message is deliberately vague to prevent a padding oracle attack.

Reply:
Makes sense.  Might be good to add a comment about this someplace in case somebody tries to improve the error messages in the future.